src/normalizer/japanese.rs

```diff
@@ -0,0 +1,194 @@
+use std::borrow::Cow;
+
+use super::{Normalizer, NormalizerOption};
+use crate::detection::{Language, Script};
+use crate::Token;
+use wana_kana::is_hiragana::*;
+use wana_kana::to_hiragana::to_hiragana_with_opt;
+use wana_kana::Options;
+
+/// Japanese specialized [`Normalizer`].
+///
+/// This Normalizer uses [`wana_kana`] internally to convert Katakana and Kanji to Hiragana
+pub struct JapaneseNormalizer;
+
+impl Normalizer for JapaneseNormalizer {
+    fn normalize<'o>(
+        &self,
+        mut token: Token<'o>,
+        options: NormalizerOption,
+    ) -> Box<dyn Iterator<Item = Token<'o>> + 'o> {
+        if is_hiragana(token.lemma()) {
+            // No need to convert
+
+            if options.create_char_map && token.char_map.is_none() {
+                let mut char_map = Vec::new();
+                for c in token.lemma().chars() {
+                    char_map.push((c.len_utf8() as u8, c.len_utf8() as u8));
+                }
+                token.char_map = Some(char_map);
+            }
+        } else {
+            // Convert Katakana to Hiragana
+            let new_lemma = to_hiragana_with_opt(
+                token.lemma(),
+                Options {
+                    pass_romaji: true, // Otherwise 'ダメ駄目だめHi' would become 'だめ駄目だめひ'
+                    ..Default::default()
+                },
+            );
+
+            if options.create_char_map && token.char_map.is_none() {
+                debug_assert!(
+                    token.lemma().len() == new_lemma.len(),
+                    concat!(
+                        r#"`to_hiragana` changed the lemma len from {} to {} but the current `char_map` computation "#,
+                        r#"expected them to be equal. If `to_hiragana` does change len of char somehow, consider "#,
+                        r#"calling `to_hiragana(char)` char by char instead of only calling `to_hiragana(lemma)` once."#
+                    ),
+                    token.lemma().len(),
+                    new_lemma.len()
+                );
+                let old_new_chars = token.lemma().chars().zip(new_lemma.chars());
```

Comment:
Unlike `to_pinyin` operates on each `char`, `to_hiragana()` can operate on the whole `lemma` at once. This provides some performance benefit because each `to_hiragana()` call has some [branching cost](https://docs.rs/wana_kana/2.1.0/src/wana_kana/to_hiragana.rs.html#24-32). However, `to_hiragana(lemma)` instead of `to_hiragana(char)` also makes the `char_map` implementation a bit more complicated, and I am not sure about the trade off here.